### ISSUE TITLE
grpc + electron = Error: package.json does not exist at /package.json

### ISSUE THREAD

**Issue by GuillaumeCisco** [grpc/grpc#13049]
### What version of gRPC and what language are you using?
 tried 1.6.6 and 1.8.0-dev (master)
 
### What operating system (Linux, Windows, …) and version?
 linux
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 node 8.4.0, python 2.17.13
 
### What did you do?
simply
`import grpc from 'grpc'` in code.
Code run on electron app with two package.json config.
 
### What did you expect to see?
 No crash
 
### What did you see instead?
`Error: package.json does not exist at /package.json`

Looks like in this code from `grpc_extension.js`:
```javascript
var binding_path = binary.find(path.resolve(path.join(__dirname, '../../../package.json')));
```
`__dirname` is `/`, instead of current directory.
Curiously if I run `__dirname` in the chrome devtools, I get the right path where the package.json is present.

**Comment by murgatroid99** [comment]
This seems like it might be a bug in electron. The `__dirname` variable is explicitly documented as referring to the full path of the directory containing the current module file.

Are you actually running this in the node 8.4.0 runtime? Is python involved here at all? What version of electron are you using?

**Comment by GuillaumeCisco** [comment]
Yes I think it is a bug in electron or something in my webpack config.
For electron, I have a special webpack config where I see:
```
    node: {
        __dirname: false,
        __filename: false,
    },
```
I have seen this issue: https://github.com/webpack/webpack/issues/1599.
Do not really know if it's related.
I copy/pasted the webpack config from a bootstrap config for running electron.

Any thoughts?

**Comment by GuillaumeCisco** [comment]
Ok just found interesting things, I disabled the `dll` part of my webpack config, and the `__dirname` changed to where my webpack file is located.
The dll webpack config for the record:
```javascript
import webpack from 'webpack';
import path from 'path';
import merge from 'webpack-merge';
import baseConfig from './base';
import {dependencies} from '../../package.json';
import rules from '../utils/rules';
import plugins from '../utils/plugins';

const dist = path.resolve(process.cwd(), 'dll');


export default merge.smart(baseConfig, {
    context: process.cwd(),
    devtool: 'eval',
    target: 'electron-renderer',
    externals: ['fsevents', 'crypto-browserify'],
    module: {
        rules: rules(),
    },
    resolve: {
        modules: [
            '../../src',
        ],
    },
    entry: {
        renderer: Object.keys(dependencies || {}).filter(o => !['babel-runtime'].includes(o)),
    },
    output: {
        library: 'renderer',
        path: dist,
        filename: '[name].dev.dll.js',
        libraryTarget: 'var',
    },
    plugins: [
        ...plugins(),
        new webpack.DllPlugin({
            path: path.join(dist, '[name].json'),
            name: '[name]',
        }),
        new webpack.LoaderOptionsPlugin({
            debug: true,
            options: {
                context: path.resolve(process.cwd(), 'src'),
                output: {
                    path: path.resolve(process.cwd(), 'dll'),
                },
            },
        }),
    ],
});
```

So If I disable it and now set the 
```javascript
node: {
        __dirname: true,
        __filename: false,
    },
```

I get the correct `__dirname` => `node_modules/grpc/src/node/src`

So I don't have the same error in the code from `grpc_extension.js`:
```javascript
exports.find = function(package_json_path,opts) {
   if (!existsSync(package_json_path)) {
        throw new Error("package.json does not exist at " + package_json_path);
   }
   var package_json = require(package_json_path);
   versioning.validate_config(package_json);
   opts = opts || {};
   if (!opts.module_root) opts.module_root = path.dirname(package_json_path);
   var meta = versioning.evaluate(package_json,opts);
   return meta.module;
};
```
For the record here is my new webpack config:
```javascript
import path from 'path';
import fs from 'fs';
import webpack from 'webpack';
import chalk from 'chalk';
import merge from 'webpack-merge';
import {spawn, execSync} from 'child_process';
import ExtractCssChunks from 'extract-css-chunks-webpack-plugin';
import HappyPack from 'happypack';


import baseConfig from './base';
import rules from '../utils/rules';
import definePlugin from '../utils/definePlugin';

const port = process.env.PORT || 1212;
const publicPath = `http://localhost:${port}/dist`;
const dll = path.resolve(process.cwd(), 'dll');
const manifest = path.resolve(dll, 'renderer.json');

const DEVELOPMENT = (['development', 'staging'].includes(process.env.NODE_ENV)),
    PRODUCTION = (['production'].includes(process.env.NODE_ENV));

/**
 * Warn if the DLL is not built
 */
// if (!(fs.existsSync(dll) && fs.existsSync(manifest))) {
//     console.log(chalk.black.bgYellow.bold(
//         'The DLL files are missing. Sit back while we build them for you with "npm run build-dll"',
//     ));
//     execSync('npm run build-dll');
// }

export default merge.smart(baseConfig, {
    devtool: 'inline-source-map',

    target: 'electron-renderer',
    entry: [
        'babel-polyfill',
        'react-hot-loader/patch',
        `webpack-dev-server/client?http://localhost:${port}/`,
        'webpack/hot/only-dev-server',
        path.join(__dirname, '../../src/client/js/index.js'),
    ],

    output: {
        publicPath,
    },

    module: {
        rules: rules('electron'),
    },

    plugins: [
        definePlugin(),
        new HappyPack({
            id: 'babel',
            loaders: [{
                path: 'babel-loader', // Options to configure babel with
                query: {
                    plugins: [
                        'universal-import',
                        'emotion',
                        'transform-runtime',
                        'lodash',
                        'date-fns',
                        'transform-class-properties',
                        'transform-es2015-classes',
                        'react-hot-loader/babel',
                    ],
                    presets: [
                        'es2015',
                        'react',
                        'stage-0',
                    ],
                },
            }],
            threads: 4,
        }),
        new ExtractCssChunks({
            filename: '[name].css',
            allChunks: false,
        }),
        // new webpack.DllReferencePlugin({
        //     context: process.cwd(),
        //     manifest: require(manifest),
        //     sourceType: 'var',
        // }),
        new webpack.HotModuleReplacementPlugin(),
        new webpack.NoEmitOnErrorsPlugin(),

        new webpack.LoaderOptionsPlugin({
            debug: true,
        }),
    ],

    node: {
        __dirname: true,
        __filename: false,
    },

    devServer: {
        port,
        publicPath,
        compress: true,
        noInfo: true,
        stats: 'errors-only',
        inline: true,
        lazy: false,
        hot: true,
        headers: {'Access-Control-Allow-Origin': '*'},
        contentBase: path.join(__dirname, '../../dist'),
        watchOptions: {
            aggregateTimeout: 300,
            ignored: /node_modules/,
            poll: 100,
        },
        historyApiFallback: {
            verbose: true,
            disableDotRule: false,
        },
        setup() {
            if (process.env.START_HOT) {
                console.log('Staring Main Process...');
                spawn(
                    'npm',
                    ['run', 'start-main-dev'],
                    {shell: true, env: process.env, stdio: 'inherit'},
                )
                    .on('close', code => process.exit(code))
                    .on('error', spawnError => console.error(spawnError));
            }
        },
    },
});
```

But now I have a new error, which is clearly weird.
It is on the require:
`var package_json = require(package_json_path);`

I get the error:
```
Uncaught (in promise) Error: Cannot find module "."
```

But if I set a breakpoint in the chrome devtools and execute it myself in the code, it works:
`Object {_from: "git+https://github.com/grpc/grpc.git#v1.6.x", _id: "grpc@1.6.6", _inBundle: false, _integrity: "sha1-kDDIjpxdCDKqV9W21yszpu+UVdo=", _location: "/grpc"…}`

So it looks now like a node/electron/grpc issue.
Do not know where to start from here, and how to fix this error.

**Comment by GuillaumeCisco** [comment]
Ok just figured out what is going on.
It's all about webpack and externals for backend node application.
This issue helped me a lot: https://github.com/chentsulin/electron-react-boilerplate/issues/1#issuecomment-107224692

The problem is that `grpc` dependency needed to be in my externals for webpack to work with, and __dirname being correct.

Got back to 
```javascript
externals: 'grpc',
node: {
        __dirname: false,
        __filename: false,
    },
```
And it worked great!

The only problem now is here in grpc_extension.js:
```javascript
var binary = require('node-pre-gyp/lib/pre-binding');
var path = require('path');
console.log(__dirname);
var binding_path = binary.find(path.resolve(path.join(__dirname, '../../../package.json')));
var binding = require(binding_path);
```

On the last line, the binding_path is:  `"/home/{linux_user}/Projects/{my_project}/node_modules/grpc/src/node/extension_binary/electron-v1.7-linux-x64/grpc_node.node"`

This path does not exist. The folder `electron-v1.7-linux-x64` does not exist, in fact I have a folder name `node-v57-linux-x64`


As this issue is now solved from its initial problem, could someone point me in the right direction for understanding what is going on now?
Thank you very much for your time, and sorry for all this displayed investigation. Hope it will help others.

**Comment by GuillaumeCisco** [comment]
Ok, as far as I understand, I need to build the binary myself as described here:
https://github.com/grpc/grpc/issues/10458
and here https://electron.atom.io/docs/tutorial/using-native-node-modules/

Si I've setup:
```
export npm_config_target=v1.7.9
export npm_config_runtime=electron
```

This is what I get:
```
> grpc@1.6.7 install /home/guillaume/Projects/morpheo/morpheo-analytics/src/electron/node_modules/grpc
> node-pre-gyp install --fallback-to-build --library=static_library

node-pre-gyp ERR! Tried to download(403): https://storage.googleapis.com/grpc-precompiled-binaries/node/grpc/v1.6.7/electron-v1.7-linux-x64.tar.gz 
node-pre-gyp ERR! Pre-built binaries not found for grpc@1.6.7 and electron@1.7.9 (electron-v1.7 ABI) (falling back to source compile with node-gyp) 
gyp WARN install got an error, rolling back install
gyp ERR! configure error 
gyp ERR! stack Error: 404 response downloading https://iojs.org/download/release/v1.7.9/iojs-v1.7.9.tar.gz
gyp ERR! stack     at Request.<anonymous> (/home/guillaume/Projects/morpheo/morpheo-analytics/node_modules/node-gyp/lib/install.js:210:14)
gyp ERR! stack     at emitOne (events.js:120:20)
gyp ERR! stack     at Request.emit (events.js:210:7)
gyp ERR! stack     at Request.onRequestResponse (/home/guillaume/Projects/morpheo/morpheo-analytics/node_modules/request/request.js:1068:10)
gyp ERR! stack     at emitOne (events.js:115:13)
gyp ERR! stack     at ClientRequest.emit (events.js:210:7)
gyp ERR! stack     at HTTPParser.parserOnIncomingClient (_http_client.js:565:21)
gyp ERR! stack     at HTTPParser.parserOnHeadersComplete (_http_common.js:116:23)
gyp ERR! stack     at TLSSocket.socketOnData (_http_client.js:454:20)
gyp ERR! stack     at emitOne (events.js:115:13)
gyp ERR! System Linux 4.10.0-37-generic
gyp ERR! command "/usr/local/bin/node" "/home/guillaume/Projects/morpheo/morpheo-analytics/node_modules/node-gyp/bin/node-gyp.js" "configure" "--fallback-to-build" "--library=static_library" "--module=/home/guillaume/Projects/morpheo/morpheo-analytics/src/electron/node_modules/grpc/src/node/extension_binary/electron-v1.7-linux-x64/grpc_node.node" "--module_name=grpc_node" "--module_path=/home/guillaume/Projects/morpheo/morpheo-analytics/src/electron/node_modules/grpc/src/node/extension_binary/electron-v1.7-linux-x64"
gyp ERR! cwd /home/guillaume/Projects/morpheo/morpheo-analytics/src/electron/node_modules/grpc
gyp ERR! node -v v8.4.0
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok 
node-pre-gyp ERR! build error 
node-pre-gyp ERR! stack Error: Failed to execute '/usr/local/bin/node /home/guillaume/Projects/morpheo/morpheo-analytics/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --library=static_library --module=/home/guillaume/Projects/morpheo/morpheo-analytics/src/electron/node_modules/grpc/src/node/extension_binary/electron-v1.7-linux-x64/grpc_node.node --module_name=grpc_node --module_path=/home/guillaume/Projects/morpheo/morpheo-analytics/src/electron/node_modules/grpc/src/node/extension_binary/electron-v1.7-linux-x64' (1)
node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/home/guillaume/Projects/morpheo/morpheo-analytics/src/electron/node_modules/node-pre-gyp/lib/util/compile.js:83:29)
node-pre-gyp ERR! stack     at emitTwo (events.js:125:13)
node-pre-gyp ERR! stack     at ChildProcess.emit (events.js:213:7)
node-pre-gyp ERR! stack     at maybeClose (internal/child_process.js:927:16)
node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:211:5)
node-pre-gyp ERR! System Linux 4.10.0-37-generic
node-pre-gyp ERR! command "/usr/local/bin/node" "/home/guillaume/Projects/morpheo/morpheo-analytics/src/electron/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build" "--library=static_library"
node-pre-gyp ERR! cwd /home/guillaume/Projects/morpheo/morpheo-analytics/src/electron/node_modules/grpc
node-pre-gyp ERR! node -v v8.4.0
node-pre-gyp ERR! node-pre-gyp -v v0.6.38
node-pre-gyp ERR! not ok 
Failed to execute '/usr/local/bin/node /home/guillaume/Projects/morpheo/morpheo-analytics/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --library=static_library --module=/home/guillaume/Projects/morpheo/morpheo-analytics/src/electron/node_modules/grpc/src/node/extension_binary/electron-v1.7-linux-x64/grpc_node.node --module_name=grpc_node --module_path=/home/guillaume/Projects/morpheo/morpheo-analytics/src/electron/node_modules/grpc/src/node/extension_binary/electron-v1.7-linux-x64' (1)
npm WARN Analytics@0.0.1 No repository field.
npm WARN Analytics@0.0.1 No license field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! grpc@1.6.7 install: `node-pre-gyp install --fallback-to-build --library=static_library`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the grpc@1.6.7 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/guillaume/.npm/_logs/2017-10-19T17_39_55_305Z-debug.log
```

Looks like now a grpc  issue again xD
What am I missing?

**Comment by GuillaumeCisco** [comment]
Ok just succeeded to build the right binary in the right folder, just exported all the variables:
```javascript
# Electron's version.
export npm_config_target=1.7.9
# The architecture of Electron, can be ia32 or x64.
export npm_config_arch=x64
export npm_config_target_arch=x64
# Download headers for Electron.
export npm_config_disturl=https://atom.io/download/electron
# Tell node-pre-gyp that we are building for Electron.
export npm_config_runtime=electron
# Tell node-pre-gyp to build module from source code.
export npm_config_build_from_source=true
```

**Comment by GuillaumeCisco** [comment]
Ok last and hope least. There one last issue...
When running my webpack config for launching the electron app, nothing is launched...
Everything seems to work correctly, I have no errors, but the app is not launched...
Like `grpc` is blocking the electron app.
I have no trace, no error.

Anybody have been in this case before?
Sorry for digressing a lot in this issue. But I think it will be helpful to describe each step for others.

**Comment by GuillaumeCisco** [comment]
Ok just found why.
I launched my electron app like:
```javascript
    mainWindow = new BrowserWindow({
        show: false,
        width: 1024,
        height: 728,
    });

    mainWindow.loadURL(`file://${__dirname}/app.html`);

    // @TODO: Use 'ready-to-show' event
    //        https://github.com/electron/electron/blob/master/docs/api/browser-window.md#using-ready-to-show-event
    mainWindow.webContents.on('did-finish-load', () => {
        if (!mainWindow) {
            throw new Error('"mainWindow" is not defined');
        }
        mainWindow.show();
        mainWindow.focus();
    });
```

If an error occurred before the `did-finish-load` event has complete, the window was never showed.
Simply passing `show` parameter from `false` to `true` helped me see the app, and detect that I did not copy my `grpc_node.node` in another folder that was needed for the electron app to run.

I succeeded all this debug thanks to chrome and its awesome debug tools with breakpoints! For people interested:
- https://blog.jetbrains.com/webstorm/2016/05/getting-started-with-electron-in-webstorm/
- https://github.com/electron/electron/issues/10445


In a nutshell, solving this issue was to put `grpc` as an external for webpack, compile my own `grpc_node.node` and placing it at the right paths.

I'll close that issue.
Thank you :)

**Comment by murgatroid99** [comment]
I would like to note that with the variables you mentioned exporting, you can actually omit the last one (`export npm_config_build_from_source=true`). If you do, it will attempt to install the precompiled `grpc_node.node` that we distribute.

**Comment by Mark086** [comment]
Issue #18267: with the same problem
